oeps/oep-0025-proc-incremental-improvements.rst

```diff
@@ -0,0 +1,247 @@
+
+* **No prerequisites.** Each INCR ticket should be doable immediately, without
+  waiting for another change to be made.
+
+* **Not urgent.** While all INCR tickets are important, they do not have
+  pressing deadlines. A new developer should have time to become familiar with
+  the contribution process and the changes to be made, without feeling
+  pressured to submit something quickly.
+
+* **Requires no additional information or clarification.** The ticket
+  description should clearly describe exactly what needs to be done.
+
+* **Specifies the required skill(s).** There should be labels on the ticket
+  listing any development skills required beyond the core ones needed to
+  submit a pull request. Typical labels include "javascript", "python", "css",
+  etc.
+
+* **Does not require knowledge of the existing Open edX code.** All INCR
+  tickets should be implementable by a developer with the appropriate
+  fundamental skills, and described assuming little or no knowledge of the
+  existing code. It can be assumed that the developer has read enough
+  documentation to start a devstack and understand the contribution process.
+
+* **Not controversial.** If the instructions in the description are followed,
+  it should be feasible to promptly review and merge the changes without
+  extended deliberation.
+
+* **Contributes to a larger goal.** Each ticket belongs to an INCR "epic"
+  ticket describing a significant improvement to the Open edX codebase which
+  depends on this small change being completed.
+
+.. _Incremental Improvements: https://openedx.atlassian.net/projects/INCR/issues/INCR-1?filter=allopenissues
+
+INCR Tickets in Context
+-----------------------
+
+Note that it is entirely possible that an epic in the INCR project may not be
+achievable purely through the completion of INCR tickets.  There may be
+changes required which require more domain knowledge of the existing code,
+can't realistically be broken down into small enough tasks, or just aren't
+understood well enough to allow writing a sufficiently detailed ticket
+description without basically completing the task.  This is all fine; INCR
+tickets should only be written for the parts of the project for which they
+are appropriate.  The rest of the work can be coordinated in a variety of
+ways; examples of such coordination are given later below.
+
+It is also not necessary to immediately detail all of the INCR tickets that
+will ultimately belong to an INCR epic when it is first created.  It is
+usually sufficient to initially create just a handful of tickets that allow
+several developers with any appropriate skill sets to contribute.  More should
+be written as the tasks are completed, so there is always something available
+to work on for those interested in contributing.  And some tasks which
+initially didn't qualify as INCR tickets because they depended on other work
+being finished first may be added to the epic as those pre-requisite tasks are
+completed.
```

Comment:
I would imagine that for some INCR epics (of all small-scoped effort), there would be some tickets that need to be implemented before follow-up tickets are.  Rather than managing multiple epics and moving tickets over to INCR once they are unblocked, is it possible to put all the work in an INCR epic, but mark unstartable work as Blocked for the time being?  The INCR filter can then keep those hidden to those looking for work.

Reply:
I suppose we could mark them as blocked.  I just want to make very sure that nobody spends an hour digging into a ticket just to realize they can't really do anything on it yet.

Reply:
Agreed.  That's why if the filter could be smart to prevent showing blocked tickets, but still allow ticket authors to keep all tickets within an epic.

This is a minor point and something we can possibly evolve with further experience.